src/main/java/io/openraven/magpie/plugins/gcp/discovery/services/ServiceDirectoryDiscovery.java

```diff
@@ -0,0 +1,129 @@
+import io.openraven.magpie.api.Session;
+import io.openraven.magpie.plugins.gcp.discovery.DiscoveryExceptions;
+import io.openraven.magpie.plugins.gcp.discovery.GCPUtils;
+import io.openraven.magpie.plugins.gcp.discovery.VersionedMagpieEnvelopeProvider;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ServiceDirectoryDiscovery implements GCPDiscovery {
+  private static final String SERVICE = "serviceDirectory";
+
+  private static final List<String> AVAILABLE_LOCATIONS = List.of(
+    "asia-east1",
+    "asia-east2",
+    "asia-northeast1",
+    "asia-northeast2",
+    "asia-northeast3",
+    "asia-south1",
+    "asia-south2",
+    "asia-southeast1",
+    "asia-southeast2",
+    "australia-southeast1",
+    "australia-southeast2",
+    "europe-central2",
+    "europe-north1",
+    "europe-west1",
+    "europe-west2",
+    "europe-west3",
+    "europe-west4",
+    "europe-west5",
+    "europe-west6",
+    "northamerica-northeast1",
+    "northamerica-northeast2",
+    "southamerica-east1",
+    "us-central1",
+    "us-central2",
+    "us-east1",
+    "us-east4",
+    "us-west1",
+    "us-west2",
+    "us-west3",
+    "us-west4");
+
+  @Override
+  public String service() {
+    return SERVICE;
+  }
+
+  @Override
+  public void discover(ObjectMapper mapper, String projectId, Session session, Emitter emitter, Logger logger) {
+    final String RESOURCE_TYPE = "GCP::ServiceDirectory::Namespace";
+
+    try (RegistrationServiceClient registrationServiceClient = RegistrationServiceClient.create()) {
+      AVAILABLE_LOCATIONS.forEach(location -> {
+        String parent = LocationName.of(projectId, location).toString();
+
+        registrationServiceClient.listNamespaces(parent).iterateAll().forEach(namespace -> {
+          var data = new MagpieResource.MagpieResourceBuilder(mapper, namespace.getName())
+            .withProjectId(projectId)
+            .withResourceType(RESOURCE_TYPE)
+            .withConfiguration(GCPUtils.asJsonNode(namespace))
+            .build();
+
+          discoverServices(registrationServiceClient, namespace.getName(), data);
+
+          emitter.emit(VersionedMagpieEnvelopeProvider.create(session, List.of(fullService() + ":namespace"), data.toJsonNode()));
+        });
+      });
+    } catch (IOException e) {
+      DiscoveryExceptions.onDiscoveryException(RESOURCE_TYPE, e);
+    }
+  }
+
+  private void discoverServices(RegistrationServiceClient registrationServiceClient,
+                                String namespaceName,
+                                MagpieResource data) {
+    final String fieldName = "services";
+
+    List<Service.Builder> services = new ArrayList<>();
+    registrationServiceClient.listServices(namespaceName).iterateAll()
+      .forEach(service -> {
+        discoverEndpoints(registrationServiceClient, service.getName(), data);
+        services.add(service.toBuilder());
```

Comment:
Discussed in private.
Envelope data reorganised to discover `Services` rather then a `Namespaces`.
Otherwise we are getting mixed data of all services and all endpoints in a part of single namespace resource, which will be difficult to map to each other.